// cassandra-orm/lib/orm/constants.js
/**
 *  Copyright 2012 Rackspace
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

var misc = require('rackspace-shared-utils/lib/misc');

/**
 * Make some basic tokens for the begin batch
 */
exports.BEGIN = 'BEGIN BATCH';

/**
 * Add a commit batch
 */
exports.COMMIT = 'APPLY BATCH;';

/**
 * Delimiter to separate columns
 */
exports.DELIM = ':';

/**
 * End token.
 */
exports.END_TOKEN = '\uffff';

/**
 * The "highest" supported version 1 time UUID there is.
 */
exports.MAX_UUID = 'fffffff1-ffff-ffff-bfff-ffffffffffff';

/**
 * The "lowest" supported version 1 time UUID there is.
 */
exports.MIN_UUID = '00000001-0000-1000-8000-000000000000';

/**
 * Maps user-friendly names of the special fields to the database names.
 * @type {Object}
 */
exports.SPECIAL_FIELDS = {
  'removed': '_removed',
  'created_at': 'created_at',
  'updated_at': 'updated_at',
  'removed_at': 'removed_at'
};

/**
 * Database names of the special fields.
 * @type {Array}
 */
exports.SPECIAL_FIELD_NAMES = Object.keys(misc.reverseObject(exports.SPECIAL_FIELDS));

/**
 * Special scope index types.
 */
exports.SCOPED_INDEX_TYPES = ['ScopedOneToMany'];


/**
 * A character used for delimiting index key from the object key in the object
 * key in the column name.
 */
exports.INDEX_KEY_DELIM = '|';
